### ISSUE TITLE
How to add authentication to the managed client

### ISSUE THREAD

**Issue by NtFreX** [aspnet/SignalR#1039]
I have a hub which requires authentication. I'm using jwt.

```
[Authorize(AuthenticationSchemes = OAuthValidationConstants.Schemes.Bearer)]
public class MyHub : Hub
{ }
```

As the access token must be sent by the query string I added a token retriever.

```
services
    .AddAuthentication()
    .AddJwtBearer(options =>
    {
        options.Events = new JwtBearerEvents();
        options.Events.OnMessageReceived += HandleOpenidConnectMessageReceived;
    }
```

```
private static Task HandleOpenidConnectMessageReceived(MessageReceivedContext ctx)
{
    if (ctx.Request.Headers.ContainsKey("Authorization"))
        return Task.CompletedTask;

    if (ctx.Request.Query.ContainsKey("token"))
        ctx.Token = ctx.Request.Query["token"];

    return Task.CompletedTask;
}
```

Now I try to use the managed client to connect to this hub.

```
public class AuthorizationQueryMessageHandler : DelegatingHandler
{
    public string AccessToken { get; set; }
        
    public AuthorizationQueryMessageHandler(string accessToken)
    {
        AccessToken = accessToken;
        InnerHandler = new HttpClientHandler();
    }

    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var newUri = QueryHelpers.AddQueryString(request.RequestUri.AbsoluteUri, "token", AccessToken);
        request.RequestUri = new Uri(newUri);
            
        return base.SendAsync(request, cancellationToken);
    }
}
```

```
var connection = new HubConnectionBuilder()
    .WithMessageHandler(new AuthorizationQueryMessageHandler(myAccessToken))
    .WithUrl("https://myPath/myHub")
    .WithConsoleLogger()
    .Build();
                
await _hubConnection.StartAsync();
```

---------------------------------------

The problem is that the client cannot connect to the hub. Maybe that not all requests have the token query string param is part of the problem?

I see the following two requests:
 - OPTIONS https://myPath/myHub?token=asde4z5zfdgsdf.sadfsdf34t.srf43rt => AuthenticationScheme: Bearer was successfully authenticated.
 - GET https://myPath/myHub?id=4727e3c3-a5b6-4865-9cc1-9bd2d27bdf79 => Authorization failed for user: (null).

I'm using the version 1.0.0-alpha2-final and build for net461.

---------------------------

How can I add jwt authentication to a singalr hub when using the managed client?

-----------------------

When I use the `TransportType.LongPolling` it works. But as I need websockets this is not a solution.

**Comment by moozzyk** [comment]
I think this is the same as: https://github.com/aspnet/SignalR/issues/686

**Comment by pcalchi** [comment]
I was able to get this working.  Some back ground information first, I'm using .net core 2 across both client and server.  Second, my tokens are being issued by an azure hosted IdentityServer4 instance and I'm using the query string to pass the tokens.  This is what my client code looks like.

` var baseUrl = $"https://xxxxx.azurewebsites.net/devicehub?t={tokenResponse.AccessToken}"; 
      var connection = new HubConnectionBuilder()
                              .WithUrl(baseUrl)
                              .WithConsoleLogger(LogLevel.Information)
                              .WithMessagePackProtocol()
                              .Build();`

Server side, since I'm using IdentityServer4, I'm using the IdentityServer4.AccessTokenValidation nuget package.  FYI, this uses the JwtBearer package under the covers.  The biggest difference between my implementation and yours is that I'm not using the OnMessageReceived event to work my magic.  

At the very top of my IApplicationBuilder chain I'm using this little piece of code to parse the token query string and pump it into request headers.  **IMPORTANT:** If you are going to use this, it's imperative that it be at the very top.

`app
.Use(async (context, next) =>
        {
          if (string.IsNullOrWhiteSpace(context.Request.Headers["Authorization"]))
          {
            try
            {
              if (context.Request.QueryString.HasValue)
              {
                var token = context.Request.QueryString.Value
                    .Split('&')
                    .SingleOrDefault(x =>
                    {
                      var i = x.IndexOf("t=", StringComparison.CurrentCultureIgnoreCase);
                      return i > -1 && i < 2;
                    })?
                    .Split('=')
                    .LastOrDefault();

                if (!string.IsNullOrWhiteSpace(token) && string.IsNullOrWhiteSpace(context.Request.Headers["Authorization"]))
                {
                  context.Request.Headers.Add("Authorization", new[] { $"Bearer {token}" });
                }
              }
            }
            catch
            {
              // if multiple headers it may throw an error.  Ignore both.
            }
          }
          await next();
        })`

**Comment by NtFreX** [comment]
@moozzyk yes I think it's the same.
@pcalchi During the first request the user gets authenticate it's just that the HttpClientHandler won't get called for further requests. What do you do when you refresh your access token? The base uri propery is private.

**Comment by pcalchi** [comment]
@NtFreX From what I understand, the token is only used to connect.  In the event of a disconnect, you would just acquire a new token to reconnect.  Messages back and forth to the server are used within the context of an existing connection.

Just as an FYI, when using bare socket connections to an Endpoint you're able to pass the token in the header.

**Comment by moozzyk** [comment]
I merged a change (0bafb304c2707739578e635bc9e00d0cd36420a7 18a65496b7adab684a9af2ac255c7128699c96ee) today that adds an API that allows to pass a JWT token when starting a connection. When using the C# client the JWT token will be passed in the header. When using the TS client the token will be passed in the header when sending an HTTP request with XmlHttpRequest (i.e. for negotiate, long polling - send/poll request, server sent events - send request) or in the query string if the underlying API does not allow setting headers (webSockets, eventSource used for the server-to-client channel in the ServerSentEvents transport) as the `signalRTokenHeader` queryString parameter. When using the TS client it is required to copy the token to the MessageReceivedContext.Token as shown in the [sample](https://github.com/aspnet/SignalR/blob/dev/samples/JwtSample/Startup.cs#L49-L62).
